_posts/2020-12-14-how-to-lock-down-csp-using-swachbuckle.md

```diff
@@ -0,0 +1,224 @@
+        {
+            // Only allow loading resources from this app by default
+            csp.AddDefaultSrc().Self();
+
+            csp.AddStyleSrc()
+               .Self()
+               // Allow nonce-enabled <style> tags
+               .WithNonce();
+
+            csp.AddScriptSrc()
+               .Self()
+               // Allow nonce-enabled <script> tags
+               .WithNonce();
+
+            [...]
+        });
+    });
+}
+```
+
+If we access our swagger-ui page again, everything seems to be working fine! 🎉
+Refreshing the page while taking a look at the HTML code, we can see that the `nonce` attribute gets generated on every request.
+
+If we look at the DevTools console again, we can see two errors remain related to loading images through the `data` protocol.
+From what I can see, they're related to the Swagger logo in the top left corner and the down-facing arrow in the definition dropdown in the top-right corner.
+
+As I believe the rest is working fine, I'll let you decide whether you want to allow loading images from `data`, which is considered insecure.
+This bit of code should do:
+
+```csharp
+public void Configure()
+{
+    app.UseSecurityHeaders(policyCollection =>
+    {
+        policyCollection.AddContentSecurityPolicy(csp =>
+        {
+            [...]
+
+            csp.AddImgSrc()
+               .Self()
+               // Allow loading images from data: links
+               .Data();
+
+            [...]
+        });
+    });
+}
+```
+
+## The `oauth2-redirect.html` file
+
+The other HTML file you might be using is the `oauth2-redirect.html` file, used by default if you decide to implement an authentication flow from swagger-ui.
+This can be beneficial as it dramatically eases the discovery of the API endpoints.
+
+At the time of writing, Swashbuckle uses swagger-ui v3.25.0.
+If we head over to the `swagger-api/swagger-ui` repository, we can find the `oauth2-redirect.html` file.
+See <https://github.com/swagger-api/swagger-ui/blob/v3.25.0/dist/oauth2-redirect.html>.
+
+CSP-wise, we have 2 issues:
+
+1. It contains an inline `<script>` tag; and
+1. It contains an inline event handler, `<body onload="run">`, which is not supported by Chrome nor the new Edge based on Chromium.
+
+Another point worth noting is that Swashbuckle doesn't expose a way to override the content of the `oauth2-redirect.html` file.
+
+The solution we came up with is to hold a slightly modified copy of the `oauth2-redirect.html` file in our application:
+
+```html
+<!-- Original file -->
+<body onload="run()">
+</body>
+<script>
+  function run () {
+      [...]
+  }
+</script>
+
+<!-- Modified copy -->
+<body>
+</body>
+<script>
+  function run () {
+      [...]
+  }
+  document.addEventListener('DOMContentLoaded', function () { run(); });
+</script>
+```
+
+By putting this copy in the `wwwroot/swagger` folder, it'll then be served by our static files middleware instead of by Swashbuckle.
```

Comment:
```suggestion
By putting this copy in the `wwwroot/swagger` folder, it'll then be served by our static files middleware instead of by the Swashbuckle SwaggerUI middleware.
```